src/natcap/invest/utils.py

```diff
@@ -597,119 +597,127 @@ def expand_path(path, base_path):
+    - Values in number, ratio, and percent columns are cast to float
+    - Values in integer columns are cast to int
+    - Values in boolean columns are cast to bool
+
+    Empty or NA cells are returned as ``numpy.nan`` (for floats) or
+    ``pandas.NA`` (for all other types).
+
+    Also sets custom defaults for some kwargs passed to ``pandas.read_csv``,
+    which you can override with kwargs:
+
+    - sep=None: lets the Python engine infer the separator
+    - engine='python': The 'python' engine supports the sep=None option.
+    - encoding='utf-8-sig': 'utf-8-sig' handles UTF-8 with or without BOM.
 
     Args:
         path (str): path to a CSV file
-        index_col (str): name of column to use as the dataframe index. If
-            ``convert_cols_to_lower``, this column name and the dataframe column names
-            will be lowercased before they are compared. If ``usecols``
-            is defined, this must be included in ``usecols``.
-        usecols (list(str)): list of column names to subset from the dataframe.
-            If ``convert_cols_to_lower``, these names and the dataframe column names
-            will be lowercased before they are compared.
-        convert_cols_to_lower (bool): if True, convert all column names to lowercase
-        convert_vals_to_lower (bool): if True, convert all table values to lowercase
-        expand_path_cols (list[string])): if provided, a list of the names of
-            columns that contain paths to expand. Any relative paths in these
-            columns will be expanded to absolute paths. It is assumed that
-            relative paths are relative to the CSV's path.
-        sep: kwarg of ``pandas.read_csv``. Defaults to None, which
-            lets the Python engine infer the separator
-        engine (str): kwarg of ``pandas.read_csv``. The 'python' engine
-            supports the sep=None option.
-        encoding (str): kwarg of ``pandas.read_csv``. Using the 'utf-8-sig'
-            encoding handles UTF-8 with or without BOM.
+        spec (dict): dictionary specifying the structure of the CSV table
         **kwargs: additional kwargs will be passed to ``pandas.read_csv``
 
     Returns:
         pandas.DataFrame with the contents of the given CSV
-
     """
+    # build up a list of regex patterns to match columns against columns from
+    # the table that match a pattern in this list (after stripping whitespace
+    # and lowercasing) will be included in the dataframe
+    patterns = []
+    for column in spec['columns']:
+        column = column.lower()
+        match = re.match(r'(.*)\[(.+)\](.*)', column)
+        if match:
+            # for column name patterns, convert it to a regex pattern
+            groups = match.groups()
+            patterns.append(f'{groups[0]}(.+){groups[2]}')
+        else:
+            # for regular column names, use the exact name as the pattern
+            patterns.append(column.replace('(', '\(').replace(')', '\)'))
```

Comment:
Needed this to handle the one case in HRA where a column name contains parentheses: `stressor buffer (meters)`. It would be nice if we could require that column names not contain any special regex characters.